app/js/PersistorHelper.js

```diff
@@ -1,24 +1,62 @@
 const crypto = require('crypto')
-const metrics = require('metrics-sharelatex')
-const meter = require('stream-meter')
 const Stream = require('stream')
 const logger = require('logger-sharelatex')
+const metrics = require('metrics-sharelatex')
 const { WriteError, ReadError, NotFoundError } = require('./Errors')
 const { promisify } = require('util')
 
 const pipeline = promisify(Stream.pipeline)
 
+// Observes data that passes through and computes some metadata for it
+// - specifically, it computes the number of bytes transferred, and optionally
+//   computes a cryptographic hash based on the 'hash' option. e.g., pass
+//   { hash: 'md5' } to compute the md5 hash of the stream
+// - if 'metric' is supplied as an option, this metric will be incremented by
+//   the number of bytes transferred
+class ObserverStream extends Stream.Transform {
+  constructor(options) {
+    super(options)
+
+    this.bytes = 0
+
+    if (options.hash) {
+      this.hash = crypto.createHash(options.hash)
```

Comment:
I am looking forward to seeing a stronger hash function used here. :+1:

Reply:
Google only give us MD5. It's fast and cheap, and works well enough for verifying integrity. Hash collisions aren't likely to be a problem for that scenario.